File: explorar_itop_tables.php
<?php
/**
 * Explorar tablas iTOP legacy para un curso.
 *
 * Uso desde el servidor:
 *   sudo -u www-data php /var/www/html/moodle/sanase/blocks/configurable_reports/explorar_itop_tables.php --courseid=67
 *
 * O copiar al directorio del plugin y ejecutar.
 */

define('CLI_SCRIPT', true);

// Detectar la raíz de Moodle (subir desde blocks/configurable_reports/).
$moodleroot = dirname(dirname(__DIR__));
if (!file_exists($moodleroot . '/config.php')) {
    // Si no estamos dentro del plugin, intentar ruta hardcoded.
    $moodleroot = '/var/www/html/moodle/sanase';
}
require($moodleroot . '/config.php');

global $DB;

// Parsear argumento --courseid
$courseid = 67;
foreach ($argv as $arg) {
    if (strpos($arg, '--courseid=') === 0) {
        $courseid = (int)substr($arg, 11);
    }
}

echo "\n=== EXPLORACIÓN iTOP LEGACY — Curso $courseid ===\n\n";

// 1. Tablas que existen
echo "--- 1. TABLAS block_adv_reports_* EXISTENTES ---\n";
$dbman = $DB->get_manager();
$tables = [
    'block_adv_reports_times',
    'block_adv_reports_usrstats',
    'block_adv_reports_values',
    'block_adv_reports_daily',
    'block_adv_reports_sco_times',
    'block_adv_reports_videoconf',
    'block_adv_reports_chours',
    'block_adv_reports_tmethod',
    'block_adv_reports_sect_times',
    'block_adv_reports_cert',
    'block_advanced_reports',
];
foreach ($tables as $t) {
    $exists = $dbman->table_exists($t) ? 'SI' : 'NO';
    echo "  $t: $exists\n";
}

// 2. Stats en usrstats
echo "\n--- 2. STATS EN block_adv_reports_usrstats (curso $courseid) ---\n";
if ($dbman->table_exists('block_adv_reports_usrstats')) {
    $sql = "SELECT stat, COUNT(*) as total, MIN(value) as min_val, MAX(value) as max_val
            FROM {block_adv_reports_usrstats}
            WHERE courseid = ?
            GROUP BY stat
            ORDER BY stat";
    $rows = $DB->get_records_sql($sql, [$courseid]);
    if ($rows) {
        printf("  %-40s %8s %15s %15s\n", 'STAT', 'TOTAL', 'MIN', 'MAX');
        printf("  %-40s %8s %15s %15s\n", str_repeat('-', 40), '--------', '---------------', '---------------');
        foreach ($rows as $r) {
            printf("  %-40s %8d %15s %15s\n", $r->stat, $r->total, $r->min_val, $r->max_val);
        }
    } else {
        echo "  (sin datos para este curso)\n";
    }
} else {
    echo "  (tabla no existe)\n";
}

// 3. Muestra usrstats
echo "\n--- 3. MUESTRA block_adv_reports_usrstats (primeros 30) ---\n";
if ($dbman->table_exists('block_adv_reports_usrstats')) {
    $rows = $DB->get_records_sql(
        "SELECT id, userid, stat, value, dim1, timecreated FROM {block_adv_reports_usrstats} WHERE courseid = ? ORDER BY userid, stat LIMIT 30",
        [$courseid]
    );
    if ($rows) {
        printf("  %-8s %-8s %-35s %-20s %-15s %-12s\n", 'ID', 'USERID', 'STAT', 'VALUE', 'DIM1', 'TIMECREATED');
        foreach ($rows as $r) {
            printf("  %-8d %-8d %-35s %-20s %-15s %-12s\n",
                $r->id, $r->userid, $r->stat,
                substr($r->value, 0, 20),
                substr($r->dim1 ?? '', 0, 15),
                $r->timecreated ? date('Y-m-d', $r->timecreated) : ''
            );
        }
    }
}

// 4. Stats en values
echo "\n--- 4. STATS EN block_adv_reports_values (curso $courseid) ---\n";
if ($dbman->table_exists('block_adv_reports_values')) {
    $sql = "SELECT stat, reportid, COUNT(*) as total
            FROM {block_adv_reports_values}
            WHERE courseid = ?
            GROUP BY stat, reportid
            ORDER BY reportid, stat";
    $rows = $DB->get_records_sql($sql, [$courseid]);
    if ($rows) {
        printf("  %-40s %-12s %8s\n", 'STAT', 'REPORTID', 'TOTAL');
        printf("  %-40s %-12s %8s\n", str_repeat('-', 40), '------------', '--------');
        foreach ($rows as $r) {
            printf("  %-40s %-12s %8d\n", $r->stat, $r->reportid, $r->total);
        }
    } else {
        echo "  (sin datos para este curso)\n";
    }
} else {
    echo "  (tabla no existe)\n";
}

// 5. Muestra values
echo "\n--- 5. MUESTRA block_adv_reports_values (primeros 30) ---\n";
if ($dbman->table_exists('block_adv_reports_values')) {
    $rows = $DB->get_records_sql(
        "SELECT id, userid, reportid, stat, value FROM {block_adv_reports_values} WHERE courseid = ? ORDER BY userid, reportid, stat LIMIT 30",
        [$courseid]
    );
    if ($rows) {
        printf("  %-8s %-8s %-12s %-35s %-20s\n", 'ID', 'USERID', 'REPORTID', 'STAT', 'VALUE');
        foreach ($rows as $r) {
            printf("  %-8d %-8d %-12s %-35s %-20s\n",
                $r->id, $r->userid, $r->reportid, $r->stat, substr($r->value, 0, 20)
            );
        }
    }
}

// 6. Times
echo "\n--- 6. MUESTRA block_adv_reports_times (primeros 20) ---\n";
if ($dbman->table_exists('block_adv_reports_times')) {
    $rows = $DB->get_records_sql(
        "SELECT t.userid, t.dedicationtime, t.graceperiods, t.timemodified
         FROM {block_adv_reports_times} t
         WHERE t.course = ?
         ORDER BY t.userid LIMIT 20",
        [$courseid]
    );
    if ($rows) {
        printf("  %-8s %-18s %-14s %-20s\n", 'USERID', 'DEDICATIONTIME', 'GRACEPERIODS', 'TIMEMODIFIED');
        foreach ($rows as $r) {
            printf("  %-8d %-18s %-14s %-20s\n",
                $r->userid, $r->dedicationtime, $r->graceperiods ?? 0,
                $r->timemodified ? date('Y-m-d H:i', $r->timemodified) : ''
            );
        }
    }
}

// 7. chours y tmethod
echo "\n--- 7. block_adv_reports_chours (curso $courseid) ---\n";
if ($dbman->table_exists('block_adv_reports_chours')) {
    $rows = $DB->get_records('block_adv_reports_chours', ['courseid' => $courseid]);
    if ($rows) {
        foreach ($rows as $r) { print_r($r); }
    } else {
        echo "  (sin datos)\n";
    }
}

echo "\n--- 8. block_adv_reports_tmethod (curso $courseid) ---\n";
if ($dbman->table_exists('block_adv_reports_tmethod')) {
    $rows = $DB->get_records('block_adv_reports_tmethod', ['courseid' => $courseid]);
    if ($rows) {
        foreach ($rows as $r) { print_r($r); }
    } else {
        echo "  (sin datos)\n";
    }
}

// 8. Daily
echo "\n--- 9. MUESTRA block_adv_reports_daily (primeros 20) ---\n";
if ($dbman->table_exists('block_adv_reports_daily')) {
    $rows = $DB->get_records_sql(
        "SELECT userid, thedate, totalseconds, totalhits FROM {block_adv_reports_daily} WHERE courseid = ? LIMIT 20",
        [$courseid]
    );
    if ($rows) {
        printf("  %-8s %-12s %-15s %-10s\n", 'USERID', 'THEDATE', 'TOTALSECONDS', 'TOTALHITS');
        foreach ($rows as $r) {
            printf("  %-8d %-12s %-15s %-10s\n", $r->userid, $r->thedate, $r->totalseconds, $r->totalhits);
        }
    }
}

// 9. SCORM
echo "\n--- 10. MUESTRA block_adv_reports_sco_times (primeros 20) ---\n";
if ($dbman->table_exists('block_adv_reports_sco_times')) {
    $rows = $DB->get_records_sql(
        "SELECT userid, scoid, attempt, dedicationtime, timemodified FROM {block_adv_reports_sco_times} WHERE course = ? LIMIT 20",
        [$courseid]
    );
    if ($rows) {
        printf("  %-8s %-8s %-8s %-18s %-20s\n", 'USERID', 'SCOID', 'ATTEMPT', 'DEDICATIONTIME', 'TIMEMODIFIED');
        foreach ($rows as $r) {
            printf("  %-8d %-8d %-8d %-18s %-20s\n",
                $r->userid, $r->scoid, $r->attempt, $r->dedicationtime,
                $r->timemodified ? date('Y-m-d H:i', $r->timemodified) : ''
            );
        }
    } else {
        echo "  (sin datos)\n";
    }
}

echo "\n=== FIN EXPLORACIÓN ===\n";

File: explorar_itop_tables.py
#!/usr/bin/env python3
"""
Explorar las tablas iTOP legacy en un servidor Moodle.

Descubre los stat names y la estructura de datos en las tablas
block_adv_reports_* para un curso determinado.

Uso:
    python explorar_itop_tables.py

Se conecta al servidor seleccionado del inventario.json y ejecuta
queries SQL via CLI de Moodle (php admin/cli/cfg.php + mysql directo).
"""

from __future__ import annotations

import json
import sys
import textwrap
from pathlib import Path

# Importar utilidades compartidas del proyecto
sys.path.insert(0, str(Path(__file__).resolve().parent))
from cr_common import (
    load_inventory,
    select_server,
    connect_ssh,
    run_remote,
    console,
)

COURSE_ID = 67  # Curso a explorar (cambiar si es necesario)

# Queries SQL para explorar las tablas iTOP
QUERIES = {
    "1_tables_exist": """
        SELECT TABLE_NAME
        FROM information_schema.TABLES
        WHERE TABLE_SCHEMA = DATABASE()
          AND TABLE_NAME LIKE 'mdl_block_adv_reports_%'
        ORDER BY TABLE_NAME;
    """,

    "2_usrstats_stats": """
        SELECT stat, COUNT(*) as total, MIN(value) as min_val, MAX(value) as max_val
        FROM mdl_block_adv_reports_usrstats
        WHERE courseid = {courseid}
        GROUP BY stat
        ORDER BY stat;
    """,

    "3_usrstats_sample": """
        SELECT userid, stat, value, dim1, timecreated
        FROM mdl_block_adv_reports_usrstats
        WHERE courseid = {courseid}
        LIMIT 30;
    """,

    "4_values_stats": """
        SELECT stat, reportid, COUNT(*) as total
        FROM mdl_block_adv_reports_values
        WHERE courseid = {courseid}
        GROUP BY stat, reportid
        ORDER BY reportid, stat;
    """,

    "5_values_sample": """
        SELECT userid, reportid, stat, value
        FROM mdl_block_adv_reports_values
        WHERE courseid = {courseid}
        LIMIT 30;
    """,

    "6_times_sample": """
        SELECT userid, dedicationtime, graceperiods, coursehours, passhours, timemodified
        FROM mdl_block_adv_reports_times t
        LEFT JOIN mdl_block_adv_reports_chours ch ON ch.courseid = t.course
        LEFT JOIN mdl_block_adv_reports_tmethod tm ON tm.courseid = t.course
        WHERE t.course = {courseid}
        LIMIT 20;
    """,

    "7_daily_sample": """
        SELECT userid, thedate, totalseconds, totalhits
        FROM mdl_block_adv_reports_daily
        WHERE courseid = {courseid}
        LIMIT 20;
    """,

    "8_scorm_sample": """
        SELECT userid, scoid, attempt, dedicationtime, timemodified
        FROM mdl_block_adv_reports_sco_times
        WHERE course = {courseid}
        LIMIT 20;
    """,

    "9_chours": """
        SELECT * FROM mdl_block_adv_reports_chours
        WHERE courseid = {courseid};
    """,

    "10_tmethod": """
        SELECT * FROM mdl_block_adv_reports_tmethod
        WHERE courseid = {courseid};
    """,

    "11_cert_sample": """
        SELECT userid, certid, timecreated
        FROM mdl_block_adv_reports_cert
        WHERE courseid = {courseid}
        LIMIT 10;
    """,

    "12_videoconf_sample": """
        SELECT userid, activityid, duration, timestart
        FROM mdl_block_adv_reports_videoconf
        WHERE courseid = {courseid}
        LIMIT 10;
    """,

    "13_sect_times_sample": """
        SELECT userid, sectionid, dedicationtime
        FROM mdl_block_adv_reports_sect_times
        WHERE courseid = {courseid}
        LIMIT 10;
    """,
}


def get_db_credentials(ssh_client, server):
    """Obtiene credenciales de la BD desde config.php de Moodle."""
    moodle_path = server["moodle_path"]
    web_user = server.get("web_user", "www-data")
    sudo_pw = server.get("sudo_password", "")

    cmd = f"sudo -u {web_user} php {moodle_path}/admin/cli/cfg.php --json 2>/dev/null || " \
          f"grep -E '\\$CFG->(dbhost|dbname|dbuser|dbpass)' {moodle_path}/config.php"

    rc, stdout, stderr = run_remote(ssh_client, cmd, step="get_db_creds",
                                     sudo_password=sudo_pw if sudo_pw else None)

    # Intentar parsear como JSON primero
    try:
        cfg = json.loads(stdout)
        return {
            "host": cfg.get("dbhost", "localhost"),
            "name": cfg.get("dbname"),
            "user": cfg.get("dbuser"),
            "pass": cfg.get("dbpass"),
        }
    except (json.JSONDecodeError, TypeError):
        pass

    # Fallback: parsear grep de config.php
    creds = {}
    for line in stdout.splitlines():
        line = line.strip().rstrip(";")
        if "dbhost" in line:
            creds["host"] = line.split("=")[-1].strip().strip("'\"")
        elif "dbname" in line:
            creds["name"] = line.split("=")[-1].strip().strip("'\"")
        elif "dbuser" in line:
            creds["user"] = line.split("=")[-1].strip().strip("'\"")
        elif "dbpass" in line:
            creds["pass"] = line.split("=")[-1].strip().strip("'\"")

    return creds


def run_sql(ssh_client, server, db_creds, query_name, query_sql):
    """Ejecuta una query SQL en el servidor remoto."""
    sql = query_sql.format(courseid=COURSE_ID).strip()
    sudo_pw = server.get("sudo_password", "")

    # Escapar la query para pasarla por bash
    escaped_sql = sql.replace("'", "'\\''")
    cmd = (
        f"mysql -h {db_creds['host']} -u {db_creds['user']} "
        f"-p'{db_creds['pass']}' {db_creds['name']} "
        f"-e '{escaped_sql}' 2>/dev/null"
    )

    try:
        rc, stdout, stderr = run_remote(ssh_client, cmd, step=query_name,
                                         sudo_password=sudo_pw if sudo_pw else None)
        return stdout
    except Exception as e:
        return f"ERROR: {e}"


def main():
    console.print(f"\n[bold cyan]Explorador de tablas iTOP legacy — Curso {COURSE_ID}[/bold cyan]\n")

    # Cargar inventario y seleccionar servidor
    inventory = load_inventory()
    server = select_server(inventory)
    if not server:
        console.print("[red]No se seleccionó servidor.[/red]")
        return

    console.print(f"\n[bold]Conectando a {server['name']}...[/bold]")
    ssh = connect_ssh(server)

    # Obtener credenciales de BD
    console.print("[dim]Obteniendo credenciales de BD...[/dim]")
    db_creds = get_db_credentials(ssh, server)
    if not db_creds.get("name"):
        console.print("[red]No se pudieron obtener las credenciales de la BD.[/red]")
        return
    console.print(f"[green]BD: {db_creds['name']} @ {db_creds['host']}[/green]\n")

    # Ejecutar cada query
    results = {}
    for qname, qsql in QUERIES.items():
        console.print(f"[cyan]Ejecutando: {qname}...[/cyan]")
        output = run_sql(ssh, server, db_creds, qname, qsql)
        results[qname] = output
        console.print(output if output.strip() else "[dim](sin resultados)[/dim]")
        console.print()

    # Guardar resultados
    output_file = Path(__file__).resolve().parent / "logs" / f"itop_explore_curso{COURSE_ID}.txt"
    with open(output_file, "w", encoding="utf-8") as f:
        f.write(f"Exploración iTOP legacy — Curso {COURSE_ID}\n")
        f.write("=" * 60 + "\n\n")
        for qname, output in results.items():
            f.write(f"--- {qname} ---\n")
            f.write(output + "\n\n")

    console.print(f"\n[bold green]Resultados guardados en: {output_file}[/bold green]")

    try:
        ssh.close()
    except Exception:
        pass


if __name__ == "__main__":
    main()
